src/engine/guidance/post_processing.cpp

```diff
@@ -287,20 +347,152 @@ std::vector<RouteStep> postProcess(std::vector<RouteStep> steps)
+                    invalidateStepAtIndex(one_back_index);
+
+                    if (TurnType::Continue == current_step.maneuver.instruction.type)
+                        steps[step_index].maneuver.instruction.type = TurnType::Turn;
+                }
+            }
+            // very short segment after turn
+            else if (TurnType::NewName == current_step.maneuver.instruction.type)
+            {
+                if (one_back_step.mode == current_step.mode)
+                {
+                    steps[step_index] =
+                        elongate(std::move(steps[step_index]), steps[one_back_index]);
+                    invalidateStepAtIndex(one_back_index);
+
+                    if (TurnType::Continue == current_step.maneuver.instruction.type)
+                    {
+                        steps[step_index].maneuver.instruction.type = TurnType::Turn;
+                    }
+                }
+            }
+            // Potential U-Turn
+            else if (bearingsAreReversed(one_back_step.maneuver.bearing_before,
+                                         current_step.maneuver.bearing_after))
+
+            {
+                // the simple case is a u-turn that changes directly into the in-name again
+                const bool direct_u_turn = steps[two_back_index].name == current_step.name;
+
+                // however, we might also deal with a dual-collapse scenario in which we have to
+                // additionall collapse a name-change as well
+                const bool u_turn_with_name_change =
+                    (step_index + 1 < steps.size()) && collapsable(current_step) &&
+                    (TurnType::NewName == steps[step_index + 1].maneuver.instruction.type) &&
+                    steps[step_index + 1].name == steps[two_back_index].name;
+
+                if (direct_u_turn || u_turn_with_name_change)
+                {
+                    steps[one_back_index] =
+                        elongate(std::move(steps[one_back_index]), steps[step_index]);
+                    invalidateStepAtIndex(step_index);
+                    if (u_turn_with_name_change)
+                    {
+                        steps[one_back_index] =
+                            elongate(std::move(steps[one_back_index]), steps[step_index + 1]);
+                        invalidateStepAtIndex(step_index + 1); // will be skipped due to the
+                                                               // continue statement at the
+                                                               // beginning of this function
+                    }
+
+                    steps[one_back_index].name = steps[two_back_index].name;
+                    steps[one_back_index].maneuver.instruction.type = TurnType::Continue;
+                    steps[one_back_index].maneuver.instruction.direction_modifier =
+                        DirectionModifier::UTurn;
+                }
+            }
+        }
+    }
```

Comment:
The function is a bit of a monster, }}}}} looks like Lisp to me :P Maybe refactor a bit.